### ISSUE TITLE
Re-organize PudlTabl output object to avoid boilerplate code

### ISSUE THREAD

**Issue by zaneselvans** [catalyst-cooperative/pudl#802]
The `pudl.output.pudltabl.PudlTabl` class which we use to provide easy access to the PUDL database (returning slightly denormalized tables) and also to many of the analysis routines which operate on the database tables is filled with boilerplate code. Most of the dozens of methods do the same thing -- call an external function that's not part of the class, and cache the resulting dataframe for later access. It seems like there should be some way to make the creation of these functions semi-automatic without so much duplicated code.

* Can we use the [functools @cached_property decorator](https://docs.python.org/3/library/functools.html#functools.cached_property) to manage the caching of the dataframes that have been pulled out of the DB, and the derived values that are involved in the MCOE calculation?
* Is there a way to automatically generate those properties just based on a list of functions that are stored within the `pudltabl` class definition, rather than needing to wrap each of them inside its own entirely boilerplate class method?

**Comment by cschloer** [comment]
OK I've looked over the code and have a few questions.

- Which functions exactly are you hoping to generate programmatically? I'm assuming [these "interim" functions](https://github.com/catalyst-cooperative/pudl/blob/main/src/pudl/output/pudltabl.py#L257) will stay as they are? And you're rather to referring to [these](https://github.com/catalyst-cooperative/pudl/blob/main/src/pudl/output/pudltabl.py#L222) [longer](https://github.com/catalyst-cooperative/pudl/blob/main/src/pudl/output/pudltabl.py#L348) [functions](https://github.com/catalyst-cooperative/pudl/blob/main/src/pudl/output/pudltabl.py#L438)? 
- For caching using cached_property, you're referring to making each dataframe itself a property of the class, rather than having this `_dfs` property that contains all of them, correct? I'm assuming from the name that the `_dfs` property is only accessed from within the class, is that assumption correct?

**Comment by cschloer** [comment]
Hi @zaneselvans ,

I've run through a first try. I haven't updated tests or the code calling this class because it will take a while and I wanted to check with you to make sure this is the direction you were imagining.

Here is the code: https://github.com/cschloer/pudl/blob/main/src/pudl/output/pudltabl.py

I've created a new function "get_data" that takes in a string (the name of the data), plus some optional arguments. The idea is that this function would be called with one of the static variables defined at the top of the file in order to ensure you don't have any typos. For example:

```
get_data(PudlTabl.pu_ferc1)
```

I didn't use cached_property, but instead kept the internal _df dictionary. I think that works fine for this usecase and cached_property doesn't really add any other functionality/advantage. There's some slightly funky/hacky code [here](https://github.com/cschloer/pudl/blob/main/src/pudl/output/pudltabl.py#L446) that I added so that not every lambda function needs an update parameter or an xargs parameters. That could be rethought but it seems to work fine as it is.

Let me know what you think and then I can make the tests + other code work with it and make a PR.
